tests/test_image_upload.py

```diff
@@ -150,3 +151,152 @@ def test_upload_image_invalid_location_header(self):
 
         with self.assertRaisesRegexp(ImageUploadException, 'S3 Invalid location header'):
             s3.upload(img_data, 'image4.png')
+
+    @responses.activate
+    def test_returns_object_url_when_aws_url_with_201_status_code(self):
+        mock_request(status=201, fixture='s3_success_response.xml',
+                     include_location=False)
+
+        signed_request = get_signed_request()
+        img_data = get_fixture('image1.png', as_binary=True)
+
+        s3 = S3ImageUploader(signed_request)
+        result = s3.upload(img_data, 'image1.png', 'png')
+
+        self.assertEqual('http://pydocx.s3.amazonaws.com/uploads/pydocx/image1.png', result)
+
+class GCSImageUploaderTestCase(TestCase):
+    @responses.activate
+    def test_content_type_not_set_when_gcs_url(self):
+        filename = 'image1.png'
+        image_data = get_fixture(filename, as_binary=True)
+        xml_body = get_fixture('gcs_success_response.xml', as_binary=True)
+
+        signed_request = get_signed_request()
+        signed_request['url'] = 'https://pydocx.storage.googleapis.com'
+
+        signed_mock = get_signed_request()
+        signed_mock.pop('url')
+        files = {"file": (filename, image_data)}
+        responses.add(responses.POST,
+                      signed_request['url'],
+                      status=201,
+                      match=[multipart_matcher(files, data=signed_mock)],
+                      content_type='multipart/form-data',
+                      body=xml_body)
+
+        S3ImageUploader(signed_request).upload(image_data, filename, 'png')
+
+    @responses.activate
+    def test_fails_when_gcs_url_with_204_status_code(self):
+        filename = 'image1.png'
+        image_data = get_fixture(filename, as_binary=True)
+
+        signed_request = get_signed_request()
+        signed_request['url'] = 'https://pydocx.storage.googleapis.com'
+
+        signed_mock = get_signed_request()
+        signed_mock.pop('url')
+        files = {"file": (filename, image_data)}
+        responses.add(responses.POST,
+                      signed_request['url'],
+                      status=204,
+                      match=[multipart_matcher(files, data=signed_mock)],
+                      content_type='multipart/form-data')
+
+        with self.assertRaisesRegexp(ImageUploadException, 'S3 Invalid location header'):
+            S3ImageUploader(signed_request).upload(image_data, filename, 'png')
```

Comment:
NON-BLOCKING: Just to check I'm understanding this test correctly, the code
```python
image_url = r.headers.get('location', None)
```
would get run because it's a 204 response, but it evaluates to `None`, and `if None` returns `false`?

Reply:
yes, that line would be hit because it's a 204 status code. 204 responses don't have bodies so we look for a `location` header. If it's not found, we return None (python's `nil`) and the conditionals fall through to an exception